pkg/resource/deploy/providers/registry.go

```diff
@@ -0,0 +1,322 @@
+		}
+
+		logging.V(7).Infof("loaded provider %v", ref)
+		r.providers[ref] = provider
+	}
+
+	return r, nil
+}
+
+func (r *Registry) GetProvider(ref Reference) (plugin.Provider, bool) {
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	logging.V(7).Infof("GetProvider(%v)", ref)
+
+	provider, ok := r.providers[ref]
+	return provider, ok
+}
+
+func (r *Registry) setProvider(ref Reference, provider plugin.Provider) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	logging.V(7).Infof("setProvider(%v)", ref)
+
+	r.providers[ref] = provider
+}
+
+func (r *Registry) deleteProvider(ref Reference) (plugin.Provider, bool) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	provider, ok := r.providers[ref]
+	if !ok {
+		return nil, false
+	}
+	delete(r.providers, ref)
+	return provider, true
+}
+
+func (r *Registry) Close() error {
+	return nil
+}
+
+func (r *Registry) Pkg() tokens.Package {
+	return "pulumi"
+}
+
+func (r *Registry) label() string {
+	return "ProviderRegistry"
+}
+
+// CheckConfig validates the configuration for this resource provider.
+func (r *Registry) CheckConfig(olds, news resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error) {
+	contract.Fail()
+	return nil, nil, errors.New("the provider registry is not configurable")
+}
+
+// DiffConfig checks what impacts a hypothetical change to this provider's configuration will have on the provider.
+func (r *Registry) DiffConfig(olds, news resource.PropertyMap) (plugin.DiffResult, error) {
+	contract.Fail()
+	return plugin.DiffResult{}, errors.New("the provider registry is not configurable")
+}
+
+func (r *Registry) Configure(props resource.PropertyMap) error {
+	contract.Fail()
+	return errors.New("the provider registry is not configurable")
+}
+
+func (r *Registry) Check(urn resource.URN, olds, news resource.PropertyMap,
+	allowUnknowns bool) (resource.PropertyMap, []plugin.CheckFailure, error) {
+
+	contract.Require(IsProviderType(urn.Type()), "urn")
+
+	label := fmt.Sprintf("%s.Check(%s)", r.label(), urn)
+	logging.V(7).Infof("%s executing (#olds=%d,#news=%d", label, len(olds), len(news))
+
+	// Parse the version from the provider properties and load the provider.
+	version, err := getProviderVersion(news)
+	if err != nil {
+		return nil, []plugin.CheckFailure{{Property: "version", Reason: err.Error()}}, nil
+	}
+	provider, err := r.host.Provider(getProviderPackage(urn.Type()), version)
+	if provider == nil {
+		return nil, nil, errors.New("could not find plugin")
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check the provider's config. If the check fails, unload the provider.
+	inputs, failures, err := provider.CheckConfig(olds, news)
+	if len(failures) != 0 || err != nil {
+		closeErr := r.host.CloseProvider(provider)
+		contract.IgnoreError(closeErr)
+		return nil, failures, err
+	}
+
+	// If we are running a preview, configure the provider now. If we are not running a preview, we will configure the
+	// provider when it is created or updated.
```

Comment:
I'm not 100% comfortable with the diverging semantics here between update and preview. To me it signals a little bit of a weakness in our model.

I get that the problem is that create and updates steps don't run on previews. I'm wondering, though, should be consistent and *always* configure the provider in Check instead of doing it just for previews? In this case, all that `Create` and `Update` do is allocate an ID and update the resource map. We're drifting away from CRUD semantics, but we already are for previews.

Reply:
I'm fine with what we have, just wondering about the implications here.